File: 0373-find-k-pairs-with-smallest-sums/0373-find-k-pairs-with-smallest-sums.cpp
class Solution {
public:
    vector<vector<int>> kSmallestPairs(vector<int>& nums1, vector<int>& nums2, int k) {
        vector<vector<int>> res_vector;
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;

        for (int x: nums1){
            pq.push({x + nums2[0], 0});
        }

        while(k-- && !pq.empty()){
            int sum = pq.top().first;
            int pos = pq.top().second;

            res_vector.push_back({sum - nums2[pos], nums2[pos]});
            pq.pop();

            if(pos + 1 < nums2.size()){
                pq.push({sum - nums2[pos] + nums2[pos + 1], pos + 1});
            }
        }

        return res_vector;
    }
};

File: 0373-find-k-pairs-with-smallest-sums/0373-find-k-pairs-with-smallest-sums.py
class Solution:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        result_vector = [] # result vector 
        pq = []            # Priority queue

        for x in nums1:
            heapq.heappush(pq, [x + nums2[0], 0])

        while k > 0 and pq:
            pair = heapq.heappop(pq)
            s, pos = pair[0], pair[1]

            result_vector.append([s - nums2[pos], nums2[pos]])

            if pos + 1 < len(nums2):
                heapq.heappush(pq, [s - nums2[pos] + nums2[pos + 1], pos + 1])

            k -= 1

        return result_vector